highest_rate.py

```diff
@@ -0,0 +1,113 @@
+# pip install kaggle-environments --upgrade -q
+from collections import defaultdict, namedtuple
+import random
+
+class Arm:
+    # An arm needs to remember the relevant history:
+    # - when my opponent selected it
+    # - when I selected it, with the reward I got
+    def __init__(self, i):
+        self.i = i
+        self.my_actions_and_rewards = {} # step number to reward
+        self.opponent_actions = set() # step number when my opponent chose it
+    def __repr__(self):
+        if len(self.my_actions_and_rewards) == 0 and len(self.opponent_actions) == 0:
+            return ''
+        return repr(dict(
+            me = list(sorted(self.my_actions_and_rewards.items())),
+            op = list(self.opponent_actions),
+            ))
+    def simple_score_from_my_pulls(self):
+        # this is really naive.
+        # Maybe better to 'discount' the old successes, as we know the arms
+        # loses 3% each time it's pull (50%-to-47%, or 50%-to-48.5%, I'm not sure)
+        how_often_have_I_pulled_this = len(self.my_actions_and_rewards)
+        reward_I_got_from_this = sum(self.my_actions_and_rewards.values())
+        return how_often_have_I_pulled_this, reward_I_got_from_this
+    def success_rate_so_far(self):
+        denominator, numerator = self.simple_score_from_my_pulls()
+        return numerator / denominator
+
+O = namedtuple('O', 'remainingOverageTime step agentIndex reward lastActions')
+
+allActions = []
+my_total_reward = 0
+my_rewards = []
+
+arms = [Arm(i) for i in range(100)]
+
+def random_agent(observation, configuration):
```

Comment:
silly question here...but how do we know what `observation` and `configuration` are?

Reply:
With the `kaggle_environments` package, we can run our agent within whichever game we are playing. This game is called `mab` (see `_run.py`). So I could run my initial agent with lots of `print` statements in order to see what is inside those objects

In the [environment rules](https://www.kaggle.com/c/santa-2020/overview/environment-rules) there is a section called `writing agents` that describes the contents of these objects. But I always like to run the code and directly see the contents :-)